filesystem/src/main/java/com/dropbox/android/external/fs3/filesystem/FileSystemImpl.kt

```diff
@@ -96,16 +95,37 @@ internal class FileSystemImpl(private val root: File) : FileSystem {
     private fun findFiles(path: String): Collection<FSFile> {
         val searchRoot = File(root, Util.simplifyPath(path))
         if (searchRoot.exists() && searchRoot.isFile) {
-            throw FileNotFoundException(format("expecting a directory at %s, instead found a file", path))
+            throw FileNotFoundException(
+                format(
+                    "expecting a directory at %s, instead found a file",
+                    path
+                )
+            )
         }
 
         val foundFiles = ArrayList<FSFile>()
         val iterator = BreadthFirstFileTreeIterator(searchRoot)
         while (iterator.hasNext()) {
             val file = iterator.next() as File?
-            foundFiles.add(files.getUnchecked(Util.simplifyPath(file!!.path
-                .replaceFirst(root.path.toRegex(), "")))!!)
+            val simplifiedPath = Util.simplifyPath(
+                file!!.path.replaceFirst(root.path.toRegex(), "")
+            )
+            foundFiles.add(files.get(simplifiedPath, fallback = FSFile(root, simplifiedPath)))
         }
         return foundFiles
     }
 }
+
+/**
+ * Cache and return a [fallback] value when no value associated with the [key] exists in the cache,
+ * otherwise return the existing value.
+ */
+private fun <Key, Value> Cache<Key, Value>.get(key: Key, fallback: Value): Value {
```

Comment:
This mimics Guava's `CacheLoader` behavior which `FileSystemImpl` was still using. Please let me know if you want this moved to the core `Cache` API.

Reply:
I think this behavior is slightly different. My understanding is that loader function was only called when the presence check fails. This allowed interactions such as cache.get(key, expensiveNetworkCall()) where the function call is lazy. Could we respect the behavior within the cache rewrite? Primarily I want to be able to call store.get(key, loader) N number of times concurrently and only have loader called once

Reply:
Sure I'll add a `get(key: Key, loader: () -> Value): Value` overload.

Reply:
I've now added cache loader support.